src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -130,7 +136,38 @@ public void Reset()
         [MethodImpl(MethodImplOptions.NoInlining)]
         private void ThrowLatchedException()
         {
-            _exceptionInfo.Throw();
+            // Only throw first exception using Edi
+            Exception exception = Interlocked.Exchange(ref _exception, _exceptionInfo.SourceException);
+
+            if (exception is null)
+            {
+                _exceptionInfo.Throw();
+            }
+
+            // Throw a new exception so as not to corrupt stack trace, as may be thrown multiple times
+            throw GetNewException();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private Exception GetNewException()
+        {
+            if (_exception is OperationCanceledException oce)
+            {
+                // Differentate OperationCanceled exceptions so Cancellation signals are propergated.
+                return new OperationCanceledException(oce.Message, oce, oce.CancellationToken);
+            }
+            else if (_exception is InvalidOperationException ioe)
+            {
+                return new InvalidOperationException(ioe.Message, ioe);
+            }
+            else if (_exception is ObjectDisposedException ode)
+            {
+                return new ObjectDisposedException(ode.Message, ode);
+            }
+            else
+            {
+                return new IOException(_exception.Message, _exception);
```

Comment:
So the issue here is that we're rethrowing the same exception instance using ExceptionDispatchInfo before another parallel thread has finished handling the exception, right?

I don't have the solution yet, but I really don't like special casing 3 exception types and wrapping everything else in an IOException.

Reply:
Yeah, so completed pipe (disconnected socket) feeding multiple Http2Streams? (might not be an accurate example..?)